### PR TITLE
Add support to **kwargs in ORTModule forward() method

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
ORTModule currently supports:
* positionals (e.g. `def forward(self, pos0, pos1, ...)`
* positionals + variable positions (e.g. `def forward(self, pos0, pos1, ..., *args)`
* positionals + keywords (e.g. `def forward(self, pos0, pos1, ..., kw0=None, kw1=None, ...)`

ORTModule currently DOES NOT support:
* Mixing variable positionals with keywords (e.g. `def forward(self, *args, kw0=None, kw1=None, ...)`
* Variable keywords (e.g. `def forward(self, ..., **kwargs)`

In this PR, support to `**kwargs` will be added, which will also support mixing `*args` with keywords

Currently, we rely on PyTorch ONNX exporter to flatten the input from (positionals, variable positionals and keywords) into a single sequential tuple of tensors. This PR will flatten the input before exporting the model. That allow us to support the aforementioned scenarios without changing the exporter, which is a third party component. A future PR will propose these changes to PyTorch

### CODE REVIEW

File: orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py [hunk 1]
```diff
@@ -214,14 +218,17 @@ def _get_exported_model(self, *inputs, **kwargs):
             _io.parse_outputs_for_onnx_export_and_extract_schema(self._original_module, inputs, kwargs)
         self._input_info.dynamic_axes.update(output_dynamic_axes)
 
+        # FlattenedModule needs _InputInfo to expand user input from *args to *args + **kwargs
+        self._flattened_module._input_info = self._input_info
```
Comment (mrry) [inline orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py:222]: Minor nit and optional: I wonder if we could refactor a little so that we don't have to mutate the flattened module. (It might make life easier in case we end up with multiple concurrent callers, or want to support caching graphs for multiple different input infos.)

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 2]
```diff
@@ -1,19 +1,50 @@
...
+            \tShape:            {self.shape}
+            \tRequire gradient: {self.require_grad_names}
+            \tDynamic axes:     {self.dynamic_axes}
+            \tSchema:           {self.schema}
+            \t#Positionals:     {self.num_positionals}
+            \tKeyword names:    {self.keyword_names}'''
+
+    def flatten(self, args, kwargs):
+        assert len(args) == self.num_positionals
+        assert set(kwargs.keys()) == set(self.keyword_names)
```
Comment (mrry) [inline orttraining/orttraining/python/training/_ortmodule_io.py:37]: Although I like how these assertions are self-documenting, they might be a little expensive (especially assuming we're checking this elsewhere), and we should probably not evaluate them at runtime.

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 3]
```diff
@@ -298,13 +323,25 @@ def parse_inputs_for_onnx_export(all_input_parameters, onnx_graph, inputs, kwarg
...
+                    print(f'{input_idx} VAR_KEYWORD {input_parameter.kind} including as keyword/kwargs {name}********************************************************************************************************************')
+                    if inp is not None and (onnx_graph is None or name in onnx_graph_input_names):
+                        if inp.requires_grad:
+                            # input_names_require_grad holds all input tensors that have requires_grad
+                            input_names_require_grad.append(name)
+                        input_names.append(name)
+                        dynamic_axes[name] = {}
+                        for dim_idx in range(len(inp.shape)):
+                            dynamic_axes[name].update({dim_idx: f'{name}_dim{dim_idx}'})
+                        input_shape.append(list(inp.size()))
```
Comment (mrry) [inline orttraining/orttraining/python/training/_ortmodule_io.py:344]: Is there a way to refactor this part of the code so that we only need to write it once for all parameter-kinds? Perhaps extract it to a local function?

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 4]
```diff
@@ -313,7 +350,9 @@ def parse_inputs_for_onnx_export(all_input_parameters, onnx_graph, inputs, kwarg
                       shape=input_shape,
                       require_grad_names=input_names_require_grad,
                       dynamic_axes=dynamic_axes,
-                      schema=schema)
+                      schema=schema,
+                      num_positionals=len(inputs),
+                      keyword_names=sorted(kwargs.keys()))
```
Comment (mrry) [inline orttraining/orttraining/python/training/_ortmodule_io.py:355]: Btw, I'm not 100% sure if the `sorted` from my example is necessary (or a good idea) :). It might be a better trade-off to rely on the natural order of Python 3.x dicts, and avoid sorting on each call.

File: orttraining/orttraining/python/training/_ortmodule_training_manager.py [hunk 5]
```diff
@@ -32,13 +32,13 @@ def forward(self, *inputs, **kwargs):
...
         if build_gradient_graph:
-            # If model was exported, then initialize the graph builder
-            self._initialize_graph_builder(training=True)
-
             # Save the onnx model if the model was exported
             if self._save_onnx:
                 onnx.save(self._onnx_model, self._save_onnx_prefix + '_exported_training_model.onnx')
 
+            # If model was exported, then initialize the graph builder
+            self._initialize_graph_builder(training=True)
```
Comment (mrry) [inline orttraining/orttraining/python/training/_ortmodule_training_manager.py:40]: Does this change do anything? I would have expected that we need to initialize the graph builder before saving a graph called "training_model" (but might be getting confused about the names...).

File: orttraining/orttraining/python/training/ortmodule.py [hunk 6]
```diff
@@ -89,50 +89,50 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
         """Override original method to delegate execution to the base module"""
 
         # Override the state_dict() method so that the state dict key names
-        # do not contain the _flattened_module._base_module prefix
-        return self._flattened_module._base_module.state_dict(
+        # do not contain the _flattened_module._original_module prefix
+        return self._flattened_module._original_module.state_dict(
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:93]: Out of curiosity, is there a reason to prefer returning `self._flattened_module._original_module....` and not `self._original_module....`?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 7]
```diff
@@ -1912,3 +1891,64 @@ def test_forward_dynamic_kwargs():
...
+            out = self.relu(out)
+            out = self.fc2(out)
+            return out
+
+    device = 'cuda'
+    N, D_in, H, D_out = 64, 784, 500, 10
+    model = KwargsNet(input_size=D_in, hidden_size=H, num_classes=D_out).to(device)
+    ort_model = ORTModule(model)
+    ort_model._execution_manager(True)._save_onnx = True
+    ort_model._execution_manager(True)._save_onnx_prefix = '/home/thiagofc/dev/github/onnxruntime/build/Linux/RelWithDebInfo/__ortmodule_args_kwargs_'
```
Comment (mrry) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1942]: (Presumably to be deleted? :))

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 8]
```diff
@@ -1912,3 +1891,64 @@ def test_forward_dynamic_kwargs():
...
+
+            if kwargs:
+                print(f'\nkwargs: {kwargs} *******************************\n')
+                if 'KwargsNet.forward.kwargs_0' in kwargs:
+                    model_input += kwargs['kwargs_0']
+
+                if 'kwargs_1' in kwargs:
+                    model_input += torch.matmul(kwargs['kwargs_0'], kwargs['kwargs_1'])
+            else:
+                print('\nNOT using KwargsNet.forward.kwargs *******************************\n')
```
Comment (mrry) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1930]: Is there a way to turn this output into a return value that you can assert on?

**Review by mrry — CHANGES_REQUESTED** [review]
Looks really good! Mostly just stylistic comments from me.

File: orttraining/orttraining/python/training/_ortmodule_io.py (hunk 4 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/_ortmodule_io.py:354]: I had a TODO on the previous iteration for this exact question, but for this spot, I think the natural ordering is fine.

File: orttraining/orttraining/python/training/_ortmodule_training_manager.py (hunk 5 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/_ortmodule_training_manager.py:40]: I just switched places with `onnx.save()`. It is quiet common that the inference model is exported but the gradient building fails. So I switched places to at least get the inference graph saved

Comment (mrry) [reply orttraining/orttraining/python/training/_ortmodule_training_manager.py:40]: In that case, can you save both, and use a different filename for the graph saved before building the training graph?

File: orttraining/orttraining/python/training/ortmodule.py (hunk 6 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py:93]: Originally introduced by https://github.com/microsoft/onnxruntime/pull/7220. Using `self._original_module` still passes `test_state_dict`, so I am making the change here (despite the unrelated

File: orttraining/orttraining/python/training/_ortmodule_training_manager.py (hunk 5 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/_ortmodule_training_manager.py:40]: We saved several versions already
XXX_exported.onnx as exported by pytorch onnx exporter
XXX_exported_training_model.onnx as the training graph
XXX_inference_optimized.onnx as the optimized inference model (from graph builder)
XXX_training_optimized.onnx  as the optimized training model (from graph builder)

Comment (mrry) [reply orttraining/orttraining/python/training/_ortmodule_training_manager.py:40]: Ah, I think I understand now. The same graph will be saved in either version (because `self._onnx_model` is the same before and after the call to `self._initialize_graph_builder()`), but now it will happen regardless of whether `self._initialize_graph_builder()` succeeds or not.

My confusion arose from the name "training_model", which doesn't sound like an inference graph :).

Makes sense now, thanks!

File: orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py (hunk 1 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py:222]: That is actually a bit tricky. I will track on a second PR. "input_info.schema" is also updated within ORTModule scope after instantiation.
Is that ok?

Comment (mrry) [reply orttraining/orttraining/python/training/_ortmodule_graph_execution_manager.py:222]: Oh yes, fine to do in a later PR!

**Comment by kit1980** [comment]
I did a test with unused inputs, and Pytorch returns 5 and ORTModule returns 3.

```
import torch
from onnxruntime.training import ORTModule
 
class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.dummy = torch.nn.Parameter(torch.FloatTensor([0]))
    def forward(self, a, b, c, d):
        return self.dummy + a + d
 
a = torch.FloatTensor([1])
b = torch.FloatTensor([2])
c = torch.FloatTensor([3])
d = torch.FloatTensor([4])
 
model = Net()
print(model(a, b, c=c, d=d))
 
model = Net()
model = ORTModule(model)
print(model(a, b, c=c, d=d))
```

**Comment by mrry** [comment]
> I did a test with unused inputs, and Pytorch returns 5 and ORTModule returns 3.
> 
> ```
> ...
> ```

@thiagocrepaldi This seems like an important case. Can you add it as a unit test (in addition to fixing it :))?

**Comment by thiagocrepaldi** [comment]
> > I did a test with unused inputs, and Pytorch returns 5 and ORTModule returns 3.
> > ```
> > ...
> > ```
> 
> @thiagocrepaldi This seems like an important case. Can you add it as a unit test (in addition to fixing it :))?

This is indeed important, but it is a pre-existing bug not related to keywords nor kwargs (the goal of this PR).
We can repro it with positional arguments only
The root cause for this scenario is that ONNX discards unused inputs and ORTModule doesn't really know about it and pass all inputs to ORT

Kwargs flattening/unflattening and names/shapes tracking are working properly as intended by the PR

I will track this separately and with priority

**Comment by mrry** [comment]
Ah, if it is pre-existing then that sounds fine to handle elsewhere. (I assume the issue crops up elsewhere, when we're building a list of inputs to pass to the `Run()` call....)

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 9]
```diff
@@ -210,51 +238,61 @@ def _populate_output_names_and_dynamic_axes(output, output_names, output_dynamic
...
-            # Copy the forward signature from the original PyTorch module.
-            self.forward = _forward.__get__(self)
-            functools.update_wrapper(self.forward.__func__, module.forward.__func__)
+        # Before `forward` is called, _ort_module must be assigned
+        # Updated input info is needed to expand args into *args, **kwargs
+        self._input_info = None
 
-    return FlattenedOutputModule(original_module)
+    def forward(self, *args):
+        # TODO: Convert *args into *args + **kwars andd call _original_module with it
```
Comment (mrry) [inline orttraining/orttraining/python/training/_ortmodule_io.py:275]: TODO is done? :)

**Review by mrry — APPROVED** [review]

File: orttraining/orttraining/python/training/_ortmodule_io.py (hunk 9 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/_ortmodule_io.py:275]: LOL

**Review by liqunfu — APPROVED** [review]
